### PR TITLE
Help should print required arguments first.

### PR DESCRIPTION
**Pull request by cmnbroad**
This is a new PR for these changes, with @lbergelson's comments addressed.

### CODE REVIEW

**Comment by akiezun** [comment]
@lbergelson please review

File: src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java [hunk 1]
```diff
@@ -177,15 +177,28 @@ public String getVersion() {
...
 
-            argumentDefinitions.stream()
-                    .filter(argumentDefinition -> printCommon || !argumentDefinition.isCommon)
-                    .forEach(argumentDefinition -> printArgumentUsage(stream, argumentDefinition));
+        // filter on common and partition on optional
+        final Map<Boolean, List<ArgumentDefinition>> argMap = argumentDefinitions.stream()
+                .filter(argumentDefinition -> printCommon || !argumentDefinition.isCommon)
+                .collect(Collectors.partitioningBy(a -> a.optional));
+
+        final List<ArgumentDefinition> reqArgs = argMap.get(false); // required args
```
Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java]: I think this crashes if it doesn't have any required args.  I don't think partitioningBy will create an empty group.

File: src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineParserTest.java [hunk 2]
```diff
@@ -78,33 +78,44 @@
...
-        clp.usage(System.out, true);
+        validateUsageOutput(clp, true); // with common args
     }
 
     @Test
     public void testUsageWithoutPositional() {
         final ArgumentsWithoutPositional fo = new ArgumentsWithoutPositional();
         final CommandLineParser clp = new CommandLineParser(fo);
-        clp.usage(System.out, false);
+        validateUsageOutput(clp, false); // without commppn args
```
Comment (lbergelson) [inline src/test/java/org/broadinstitute/hellbender/cmdline/CommandLineParserTest.java]: typo in commppn

File: src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java (hunk 1 above)
Comment (akiezun) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java]: a test will clarify this question and make sure we do the right thing

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java]: Wait, no it doesn't, I'm mixing java and python.  Carry on...

Comment (lbergelson) [inline src/main/java/org/broadinstitute/hellbender/cmdline/CommandLineParser.java]: A test case with no options wouldn't be bad, but it definitely shouldn't crash.  Temporary confusion with `Optional.get` and python maps.

**Comment by lbergelson** [comment]
@cmnbroad Back to you.  Looks good.  Much happier with this.  Could you fix the typo and add a test case with no options just to make sure nothing weird happens.

**Comment by lbergelson** [comment]
@cmnbroad Not sure if you know, but you don't have to close and reopen pull requests to update them.  Just push changes to your branch and it will auto update.  You can force push if necessary.

**Comment by cmnbroad** [comment]
Only change here from the last round was to add some new tests.

**Comment by akiezun** [comment]
:+1: rebase and merge at will. good job on the tests

**Pull request merged**